build: add scripts to work with patches for dependencies

RaduBerinde opened the pull request:
This commit adds three scripts:
 - patch-prepare-repo.sh prepares a clone of the dependency by
   generating the `WORKSPACE` and `BUILD.bazel` files (using gazelle).
 - patch-apply.sh: applies an exiting patch to the repository.
 - patch-gen.sh: generates the patch from the current diff in the
   repository.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/133523)
<!-- Reviewable:end -->

rickystewart reviewed (commented):
The `build/patches` directory is supposed to be for patches. Can you please put these in a new directory? For example, `build/scripts`.

RaduBerinde commented:
Done.

rickystewart commented on build/scripts/patch-prepare-repo.sh:
```diff
@@ -0,0 +1,77 @@
+fi
+
+echo "Running gazelle update ${@:3} --go_prefix=\"$IMPORT_PATH\" --repo_root=."
+go run github.com/bazelbuild/bazel-gazelle/cmd/gazelle@v0.37.0 update ${@:3} --go_prefix="$IMPORT_PATH" --repo_root=.
```
Did you try using `bazel run @bazel_gazelle//cmd/gazelle --run_under="cd $PWD && "` (or something to that effect)? This would mean that we don't have to keep this script in sync with the version of `gazelle` used in `WORKSPACE`.

RaduBerinde replied:
Good idea, I'll try it.

RaduBerinde replied:
I'm having some issues making it work. The command succeeds but it does nothing.

I'm doing
```
  bazel run @bazel_gazelle//cmd/gazelle --run_under="cd $REPO_DIR && " -- update ${@:3} --go_prefix="$IMPORT_PATH" --repo_root=.
```

Output:
```
+ bazel run @bazel_gazelle//cmd/gazelle '--run_under=cd /Users/radu/pebble && ' -- update --go_prefix=github.com/cockroachdb/pebble --repo_root=.
INFO: Invocation ID: d001ebae-f750-4880-bc4c-18010a3bcd26
INFO: Analyzed target @@bazel_gazelle//cmd/gazelle:gazelle (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target @@bazel_gazelle//cmd/gazelle:gazelle up-to-date:
  _bazel/bin/external/bazel_gazelle/cmd/gazelle/gazelle_/gazelle
INFO: Elapsed time: 0.166s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Running command line: /bin/bash -c 'cd /Users/radu/pebble &&  _bazel/bin/external/bazel_gazelle/cmd/gazelle/gazelle_/gazelle update '\''--go_prefix=github.com/cockroachdb/pebble'\'' '\''--repo_root=.'\'''
```

This looks like it wouldn't work because `_bazel/bin/external/bazel_gazelle/cmd/gazelle/gazelle_/gazelle` is no longer valid after `cd`.

I see `run_under` is used similarly in other scripts, not sure why it doesn't work in this case.

I couldn't find a way to make gazelle run correctly with the crdb root as the cwd.

RaduBerinde replied:
It does run the command (it shows the help if I pass `-help` at the end). But it just doesn't do anything, still trying to figure out why.

RaduBerinde replied:
Not sure what's with the weird escaping of the update arguments. It works if I manually run (from the other repo)
```
~/roach/_bazel/bin/external/bazel_gazelle/cmd/gazelle/gazelle_/gazelle update --go_prefix=github.com/cockroachdb/pebble --repo_root=.
```

RaduBerinde replied:
Still not sure why the run_under stuff didn't work but I made it work like this:
```
bazel run @bazel_gazelle//cmd/gazelle -- update ${@:3} --go_prefix="$IMPORT_PATH" --repo_root=$REPO_DIR $REPO_DIR
```

rickystewart replied:
Nice.

rickystewart commented on build/scripts/patch-prepare-repo.sh:
```diff
@@ -0,0 +1,77 @@
+
+  TEMP_FILE=$(mktemp)
+  cp .gitignore $TEMP_FILE
+  sed -E '\#^[::space::]*(/WORKSPACE|BUILD.bazel)[::space::]*$#d' $TEMP_FILE >.gitignore
```
Nit: can use `sed -i` to update in-place instead of manually creating a temporary file

RaduBerinde replied:
I couldn't find a portable way to do it, sed -i doesn't work on Mac the same way.

rickystewart commented on build/scripts/patch-gen.sh:
```diff
@@ -0,0 +1,56 @@
+# Use of this software is governed by the CockroachDB Software License
+# included in the /LICENSE file.
+
+set -euo pipefail
```
The script is fine. It does seem like overkill to essentially do `git diff > file` though.

RaduBerinde replied:
Right, I added it just for "symmetry" with the apply one. It does save you a little bit in that you don't have to worry about where the patch goes.

RaduBerinde commented on build/patches/patch-prepare-repo.sh:
```diff
@@ -0,0 +1,77 @@
+	echo "workspace(name = \"$REPO_NAME\")" > WORKSPACE
+fi
+
+echo "Running gazelle update ${@:3} --go_prefix=\"$IMPORT_PATH\" --repo_root=."
```
Done.

RaduBerinde commented:
TFTR!

I will merge later once I get a chance to use them more.

RaduBerinde committed:
build: add scripts to work with patches for dependencies

This commit adds three scripts:
 - patch-prepare-repo.sh prepares a clone of the dependency by
   generating the `WORKSPACE` and `BUILD.bazel` files (using gazelle).
 - patch-apply.sh: applies an exiting patch to the repository.
 - patch-gen.sh: generates the patch from the current diff in the
   repository.

Epic: none
Release note: None

RaduBerinde commented:
I used these a bit and they've been helpful. The only issue was a slightly different `BUILD.bazel` file because of different gazelle directives. I added a mention in `patch-gen.sh` to help the user get the actual file from bazel.

bors r+

RaduBerinde commented:
@rickystewart One difference between the bazel files generated by the build system and those generated by the gazelle call in my script is that the former has something like this at the end:
```
alias(
    name = "go_default_library",
    actual = ":invariants",
    visibility = ["//:__subpackages__"],
)
```

This leads to patches that sometimes don't apply. Do you know where this difference come from? Is there a flag I can pass to gazelle to generate that bit as well?

rickystewart commented:
I would [check out the](https://github.com/bazel-contrib/bazel-gazelle) `README`. This is related to the `go_naming_convention` configuration. Specifically in this context, `gazelle` is inferring the default behavior should be `import_alias`. You should set it manually to `import`.
